src/transformStateWithClones.js

```diff
@@ -6,8 +6,40 @@
  *
  * @return {Object[]}
  */
+
+const ACTION_ADD_PROPERTIES = 'addProperties';
+const ACTION_REMOVE_PROPERTIES = 'removeProperties';
+const ACTION_CLEAR = 'clear';
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopy = { ...state };
+  const actionsOnState = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case ACTION_ADD_PROPERTIES:
+        Object.assign(stateCopy, action.extraData);
+        break;
+
+      case ACTION_REMOVE_PROPERTIES:
+        action.keysToRemove.forEach((key) => {
+          delete stateCopy[key];
+        });
+        break;
+      case ACTION_CLEAR:
+        for (const key in stateCopy) {
+          delete stateCopy[key];
+        }
+        break;
+
+      default:
+        throw new Error('unsupported value');
+    }
+
+    actionsOnState.push(Object.assign({}, stateCopy));
+  }
```

Comment:
my suggestion would be to use the spread operator when it is possible (i.e. wherever you're not assigning it to the function parameter), I find this more readable:
```suggestion
    actionsOnState.push({ ...stateCopy });
```